// assets-manager/dist/Registry/Custom.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _assert = require('assert');

var _assert2 = _interopRequireDefault(_assert);

var _Registry2 = require('../Registry');

var _Registry3 = _interopRequireDefault(_Registry2);

var _file = require('../file');

var _file2 = _interopRequireDefault(_file);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _configure = require('../configure');

var _configure2 = _interopRequireDefault(_configure);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Custom = function (_Registry) {
  _inherits(Custom, _Registry);

  function Custom(name) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    _classCallCheck(this, Custom);

    var _this = _possibleConstructorReturn(this, (Custom.__proto__ || Object.getPrototypeOf(Custom)).call(this, options));

    Object.assign(_this.options, {
      json: 'package.json',
      dir: _configure2.default.get('registry.' + name + '.dir', null),
      overrides: {}
    }, options);

    (0, _assert2.default)(_this.options.dir !== null, 'options.dir must not be empty');

    _this.name = name;
    return _this;
  }

  _createClass(Custom, [{
    key: 'getPackageInfo',
    value: function getPackageInfo(packageName, key) {
      var packagePath = this.getRelativePackagePath(packageName);

      if (!_file2.default.isDirectory(packagePath)) {
        return {};
      }

      var infoPath = _path2.default.join(packagePath, this.options.json);
      var info = {};

      if (_file2.default.exists(infoPath)) {
        info = _file2.default.readJSON(infoPath);
      } else {
        info = {
          name: packageName
        };
      }

      if (typeof key !== 'undefined') {
        if (Object.hasOwnProperty.call(info, key)) {
          return info[key];
        }
        return undefined;
      }
      return info;
    }
  }, {
    key: 'getPackageInstallCmd',
    value: function getPackageInstallCmd(packageName) {
      var packagePath = this.getRelativePackagePath(packageName);
      return 'check ' + packagePath + ' path';
    }
  }]);

  return Custom;
}(_Registry3.default);

exports.default = Custom;
module.exports = exports['default'];
