// assets-manager/dist/Registry/Npm.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _Registry2 = require('../Registry');

var _Registry3 = _interopRequireDefault(_Registry2);

var _file = require('../file');

var _file2 = _interopRequireDefault(_file);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _configure = require('../configure');

var _configure2 = _interopRequireDefault(_configure);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Npm = function (_Registry) {
  _inherits(Npm, _Registry);

  function Npm() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    _classCallCheck(this, Npm);

    var _this = _possibleConstructorReturn(this, (Npm.__proto__ || Object.getPrototypeOf(Npm)).call(this, options));

    Object.assign(_this.options, {
      json: 'package.json',
      overrides: {},
      dir: _configure2.default.get('registry.npm.dir', 'node_modules')
    }, options);
    return _this;
  }

  _createClass(Npm, [{
    key: 'getPackageMainFiles',
    value: function getPackageMainFiles(packageName) {
      var info = this.getPackageInfo(packageName);

      if (info && info.main && typeof info.main === 'string') {
        return [info.main];
      }

      return [];
    }
  }, {
    key: 'getPackageInfo',
    value: function getPackageInfo(packageName, key) {
      var infoPath = _path2.default.join(this.getRelativePackagePath(packageName), this.options.json);
      var info = {};

      if (_file2.default.exists(infoPath)) {
        info = _file2.default.readJSON(infoPath);
      }

      if (typeof key !== 'undefined') {
        if (Object.hasOwnProperty.call(info, key)) {
          return info[key];
        }
        return undefined;
      }
      return info;
    }
  }, {
    key: 'getPackageInstallCmd',
    value: function getPackageInstallCmd(packageName) {
      return '\'npm install ' + packageName + ' --save\'';
    }
  }]);

  return Npm;
}(_Registry3.default);

exports.default = Npm;
module.exports = exports['default'];
